channeldb/payments.go

```diff
@@ -66,265 +116,334 @@ func (ps *PaymentStatus) FromBytes(status []byte) error {
+	// made.
+	//
+	// NOTE: Can be nil if no attempt is yet made.
+	*PaymentAttemptInfo
 
-		return payments.Put(paymentIDBytes, paymentBytes)
-	})
+	// PaymentPreimage is the preimage of a successful payment. This serves
+	// as a proof of payment. It will be all zeroes for non-settled
+	// payments.
+	PaymentPreimage lntypes.Preimage
 }
 
-// FetchAllPayments returns all outgoing payments in DB.
-func (db *DB) FetchAllPayments() ([]*OutgoingPayment, error) {
-	var payments []*OutgoingPayment
+// FetchPayments returns all sent payments found in the DB.
+func (db *DB) FetchPayments() ([]*Payment, error) {
+	var payments []*Payment
 
 	err := db.View(func(tx *bbolt.Tx) error {
-		bucket := tx.Bucket(paymentBucket)
-		if bucket == nil {
-			return ErrNoPaymentsCreated
+		paymentsBucket := tx.Bucket(paymentsRootBucket)
+		if paymentsBucket == nil {
+			return nil
 		}
 
-		return bucket.ForEach(func(k, v []byte) error {
-			// If the value is nil, then we ignore it as it may be
-			// a sub-bucket.
-			if v == nil {
-				return nil
+		return paymentsBucket.ForEach(func(k, v []byte) error {
+			bucket := paymentsBucket.Bucket(k)
+			if bucket == nil {
+				// We only expect sub-buckets to be found in
+				// this top-level bucket.
+				return fmt.Errorf("non bucket element")
 			}
 
-			r := bytes.NewReader(v)
-			payment, err := deserializeOutgoingPayment(r)
+			var (
+				err error
+				p   = &Payment{}
+			)
+
+			seqBytes := bucket.Get(paymentSequenceKey)
+			if seqBytes == nil {
+				return fmt.Errorf("sequence number not found")
+			}
+
+			p.sequenceNum = binary.BigEndian.Uint64(seqBytes)
+
+			// Get the payment status.
+			p.Status = fetchPaymentStatus(bucket)
+
+			// Get the PaymentCreationInfo.
+			b := bucket.Get(paymentCreationInfoKey)
+			if b == nil {
+				return fmt.Errorf("creation info not found")
+			}
+
+			r := bytes.NewReader(b)
+			p.PaymentCreationInfo, err = deserializePaymentCreationInfo(r)
 			if err != nil {
 				return err
+
 			}
 
-			payments = append(payments, payment)
+			// Get the PaymentAttemptInfo. This can be unset.
+			b = bucket.Get(paymentAttemptInfoKey)
+			if b != nil {
+				r = bytes.NewReader(b)
+				p.PaymentAttemptInfo, err = deserializePaymentAttemptInfo(r)
+				if err != nil {
+					return err
+				}
+			}
+
+			// Get the payment preimage. This is only found for
+			// completed payments.
+			b = bucket.Get(paymentSettleInfoKey)
+			if b != nil {
+				copy(p.PaymentPreimage[:], b[:])
+			}
+
+			payments = append(payments, p)
 			return nil
 		})
 	})
 	if err != nil {
 		return nil, err
 	}
 
+	// Before returning, sort the payments by their sequence number.
+	sort.Slice(payments, func(i, j int) bool {
+		return payments[i].sequenceNum < payments[j].sequenceNum
+	})
+
 	return payments, nil
 }
 
-// DeleteAllPayments deletes all payments from DB.
-func (db *DB) DeleteAllPayments() error {
+// DeletePayments deletes all payments from the DB.
+func (db *DB) DeletePayments() error {
```

Comment:
Move to control tower?

Reply:
Added here since the `rpcserver` don't have access to the control tower.